[DDP Comm Hook] Use a single tensor instead of a tensor list as the comm hook result

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62074
* #62073

Since SPMD mode is retired, the comm hook result will always be a single tensor.

This can improve comm hook developer experience, as no need to add an extra `[0]` to the precursor future result.

#Closes: https://github.com/pytorch/pytorch/issues/61914

Differential Revision: [D29864732](https://our.internmc.facebook.com/intern/diff/D29864732/)

Commit:
[DDP Comm Hook] Use a single tensor instead of a tensor list as the comm hook result

Since SPMD mode is retired, the comm hook result will always be a single tensor.

This can improve comm hook developer experience, as no need to add an extra `[0]` to the precursor future result.

#Closes: https://github.com/pytorch/pytorch/issues/61914

Differential Revision: [D29864732](https://our.internmc.facebook.com/intern/diff/D29864732/)

[ghstack-poisoned]

Commit:
Update on "[DDP Comm Hook] Use a single tensor instead of a tensor list as the comm hook result"


Since SPMD mode is retired, the comm hook result will always be a single tensor.

This can improve comm hook developer experience, as no need to add an extra `[0]` to the precursor future result.

#Closes: https://github.com/pytorch/pytorch/issues/61914

Differential Revision: [D29864732](https://our.internmc.facebook.com/intern/diff/D29864732/)

[ghstack-poisoned]

Commit:
Update on "[DDP Comm Hook] Use a single tensor instead of a tensor list as the comm hook result"


Since SPMD mode is retired, the comm hook result will always be a single tensor.

This can improve comm hook developer experience, as no need to add an extra `[0]` to the precursor future result.

#Closes: https://github.com/pytorch/pytorch/issues/61914

Differential Revision: [D29864732](https://our.internmc.facebook.com/intern/diff/D29864732/)

[ghstack-poisoned]

Commit:
Update on "[DDP Comm Hook] Use a single tensor instead of a tensor list as the comm hook result"


Since SPMD mode is retired, the comm hook result will always be a single tensor.

This can improve comm hook developer experience, as no need to add an extra `[0]` to the precursor future result.

#Closes: https://github.com/pytorch/pytorch/issues/61914

Differential Revision: [D29864732](https://our.internmc.facebook.com/intern/diff/D29864732/)

[ghstack-poisoned]

Commit:
Update on "[DDP Comm Hook] Use a single tensor instead of a tensor list as the comm hook result"


Since SPMD mode is retired, the comm hook result will always be a single tensor.

This can improve comm hook developer experience, as no need to add an extra `[0]` to the precursor future result.

#Closes: https://github.com/pytorch/pytorch/issues/61914

Differential Revision: [D29864732](https://our.internmc.facebook.com/intern/diff/D29864732/)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/c10d/comm.hpp:
```diff
@@ -81,38 +81,38 @@ class TORCH_API GradBucket {
   }
 
-  return result.toTensorVector();
+  return result.toTensorVector()[0];
```
Can cpp comm. hook somehow still return a tensor list? If not, should we be able to remove this block entirely and expect `result` to always be a tensor?

SciPioneer replied:
Good observation! This can be refactored in the next PR. Currently only `allreduce_hook` which directly calls `pg.allreduce` still returns a tensor list. I need to confirm with @lw if it's OK to do `dist.all_reduce(tensor).get_future().wait()[0]`.

rohan-varma commented on torch/csrc/distributed/c10d/python_comm_hook.cpp:
```diff
@@ -39,22 +39,21 @@ c10::intrusive_ptr<c10::ivalue::Future> PythonCommHook::runHook(
-  return result.toTensorVector();
+  // Only if the hook is an `allreduce_hook`, as the vanilla allreduce returns a
+  // tensor vector.
+  return result.toTensorVector()[0];
```
Should we modify allreduce hook to return a single tensor?

rohan-varma replied:
I see that `get_future()` returns a future[list[tensor]], is it worth it to refactor the hook to do something like

```
return dist.all_reduce(tensor).get_future().then(lambda li: return lu[0])
```

SciPioneer replied:
I am thinking of `dist.all_reduce(tensor).get_future().wait()[0]` instead. @lw Could you please confirm that this will still remain the asynchronous property?

rohan-varma replied:
IIUC `return dist.all_reduce(tensor).get_future().wait()[0]` won't work because it is a single tensor, but the contract for the comm. hook is that it needs to return a future to a single tensor?

SciPioneer replied:
Will update the code with `dist.all_reduce(tensor).get_future().then(lambda li: return lu[0])` in another PR.

rohan-varma reviewed (approved):
LGTM, thanks for improving the API!

Is it correct to say that user can still write a comm. hook that returns Future[list[single tensor]] and we won't throw errors? It seems that is how the default allreduce hook is still implemented and working. Because of this, I'm assuming its non-BC breaking and we don't have to fix any internal callsites?

SciPioneer commented:
> LGTM, thanks for improving the API!
> 
> Is it correct to say that user can still write a comm. hook that returns Future[list[single tensor]] and we won't throw errors? It seems that is how the default allreduce hook is still implemented and working. Because of this, I'm assuming its non-BC breaking and we don't have to fix any internal callsites?

Currently only `allreduce_hook` that directly calls `pg.allreduce` returns Future[list[single tensor]. All the other hooks that require `then` will essentially be a `pyobject` in C++ code, and it will be parsed as a single tensor. I am thinking of returning  a single tensor for `allreduce_hook` as well, for the consistency.

Commit:
Update on "[DDP Comm Hook] Use a single tensor instead of a tensor list as the comm hook result"


Since SPMD mode is retired, the comm hook result will always be a single tensor.

This can improve comm hook developer experience, as no need to add an extra `[0]` to the precursor future result.

#Closes: https://github.com/pytorch/pytorch/issues/61914

Differential Revision: [D29864732](https://our.internmc.facebook.com/intern/diff/D29864732/)

[ghstack-poisoned]

lw commented:
> I am thinking of dist.all_reduce(tensor).get_future().wait()[0] instead. @lw Could you please confirm that this will still remain the asynchronous property?

Well that expression is not asynchronous, since it does a wait and thus blocks. Though I suspect I am missing something, since that expression was already present before this PR, rather than being introduced here. Generally speaking, eventually at some point we must wait and block, because we must "convert" a Future to a Tensor, the question is finding the _best_ place to do so (e.g., if we have multiple futures, we should wait them all at once, together).

SciPioneer commented:
> > I am thinking of dist.all_reduce(tensor).get_future().wait()[0] instead. @lw Could you please confirm that this will still remain the asynchronous property?
> 
> Well that expression is not asynchronous, since it does a wait and thus blocks. Though I suspect I am missing something, since that expression was already present before this PR, rather than being introduced here. Generally speaking, eventually at some point we must wait and block, because we must "convert" a Future to a Tensor, the question is finding the _best_ place to do so (e.g., if we have multiple futures, we should wait them all at once, together).

Thanks for the confirmation! Then we shouldn't do this change. This PR only changes a tensor list to a single tensor for the subsequent callbacks in `then()`, but the first allreduce should still output a tensor list.
